store/tikv/latch/latch.go

```diff
@@ -33,120 +33,161 @@ type Latch struct {
+)
+
 // NewLock creates a new lock.
-func NewLock(startTS uint64, requiredSlots []int) Lock {
-	return Lock{
+func NewLock(startTS uint64, requiredSlots []int) *Lock {
+	return &Lock{
 		requiredSlots: requiredSlots,
 		acquiredCount: 0,
-		isWaiting:     false,
 		startTS:       startTS,
 	}
 }
 
+// IsStale returns whether the status is stale.
+func (l *Lock) IsStale() bool {
+	return l.status == acquireStale
+}
+
 // Latches which are used for concurrency control.
 // Each latch is indexed by a slot's ID, hence the term latch and slot are used in interchangeable,
 // but conceptually a latch is a queue, and a slot is an index to the queue
 type Latches struct {
-	slots []Latch
-	// The waiting queue for each slot(slotID => slice of startTS).
-	waitingQueues map[int][]uint64
+	slots []latch
+	// The waiting queue for each slot(slotID => slice of Lock).
+	waitingQueues map[int][]*Lock
 	sync.RWMutex
 }
 
 // NewLatches create a Latches with fixed length,
 // the size will be rounded up to the power of 2.
 func NewLatches(size int) *Latches {
 	powerOfTwoSize := 1 << uint32(bits.Len32(uint32(size-1)))
-	slots := make([]Latch, powerOfTwoSize)
+	slots := make([]latch, powerOfTwoSize)
 	return &Latches{
 		slots:         slots,
-		waitingQueues: make(map[int][]uint64),
+		waitingQueues: make(map[int][]*Lock),
 	}
 }
 
 // GenLock generates Lock for the transaction with startTS and keys.
-func (latches *Latches) GenLock(startTS uint64, keys [][]byte) Lock {
+func (latches *Latches) GenLock(startTS uint64, keys [][]byte) *Lock {
+	return NewLock(startTS, latches.genSlotIDs(keys))
+}
+
+func (latches *Latches) genSlotIDs(keys [][]byte) []int {
 	slots := make([]int, 0, len(keys))
 	for _, key := range keys {
 		slots = append(slots, latches.slotID(key))
 	}
 	sort.Ints(slots)
 	if len(slots) <= 1 {
-		return NewLock(startTS, slots)
+		return slots
 	}
 	dedup := slots[:1]
 	for i := 1; i < len(slots); i++ {
 		if slots[i] != slots[i-1] {
 			dedup = append(dedup, slots[i])
 		}
 	}
-	return NewLock(startTS, dedup)
+	return dedup
 }
 
 // slotID return slotID for current key.
 func (latches *Latches) slotID(key []byte) int {
 	return int(murmur3.Sum32(key)) & (len(latches.slots) - 1)
 }
 
-// Acquire tries to acquire the lock for a transaction.
-// It returns with stale = true when the transaction is stale(
-// when the lock.startTS is smaller than any key's last commitTS).
-func (latches *Latches) Acquire(lock *Lock) (success, stale bool) {
+// acquire tries to acquire the lock for a transaction.
+func (latches *Latches) acquire(lock *Lock) acquireResult {
+
 	for lock.acquiredCount < len(lock.requiredSlots) {
 		slotID := lock.requiredSlots[lock.acquiredCount]
-		success, stale = latches.acquireSlot(slotID, lock.startTS)
-		if success {
+		status := latches.acquireSlot(slotID, lock)
+		// The current slotID is acquired when previous status is locked while
+		// current status is stale.
+		if status == acquireStale && lock.status == acquireLocked {
```

Comment:
We can return immediately when status is stale.
And when status is stale, acquireSlot would not put this lock into the latch.

Reply:
So should we use `isWaiting` again?

Reply:
In `release`, we need to release lock according to lock's status:
1. For status is a success, the number of latches to release is acquiredCount
2. For status is locked, the number of latches to release is acquiredCount +1 since the last latch is locked.
3. For status is stale, we only know the last latch lock is stale, while do not know whether the last latch should be released since a latch could be stale in the following situations:
    1. stale in the first call of `acquireSlot`, it needn't be released since it never comes into the waiting queue.
    2. stale in the second `acquireSlot`, which means it's already in the waiting queue when stale happens, in this situation it should be released.

So we could use `isWaiting` or increased acquiredCount to mark the second situation. So maybe removing `isWaiting`  easier to understand?